File: implementations/javascript/bin/ff3-benchmark.js
#!/usr/bin/env node

/**
 * FF3 Performance Benchmark Tool
 *
 * Measures FF3 encryption/decryption performance across different string lengths.
 *
 * ⚠️ FF3 was withdrawn by NIST due to security vulnerabilities.
 * This implementation is for educational and research purposes only.
 */

import fs from 'fs';
import os from 'os';
import {
    digits,
    hexLower,
    base36Lower,
    base62,
    hexToBytes,
    VERSION
} from '../lib/index.js';

/**
 * Default configuration
 */
const DEFAULT_CONFIG = {
    alphabet: 'digits',
    radix: 10,
    lengths: [9, 12, 16],
    cases: ['enc', 'dec', 'roundtrip'],
    iterations: 100000,
    warmup: 10000,
    key: 'EF4359D8D580AA4F7F036D6F04FC6A94',
    tweak: 'D8E7920AFA330A73',
    seed: 42,
    verbose: false,
    jsonOut: null
};

/**
 * Parse CSV string to integer array
 */
function parseCsvInts(csv) {
    return csv.split(',').map(s => parseInt(s.trim(), 10));
}

/**
 * Parse CSV string to string array
 */
function parseCsvStrings(csv) {
    return csv.split(',').map(s => s.trim());
}

/**
 * Simple seeded random number generator (LCG)
 */
function createSeededRandom(seed) {
    let state = seed;
    return function() {
        state = (state * 1103515245 + 12345) & 0x7fffffff;
        return state / 0x7fffffff;
    };
}

/**
 * Generate deterministic test inputs with seed
 */
function generateInputs(alphabet, length, count, seed) {
    const random = createSeededRandom(seed);
    const inputs = [];
    const chars = alphabet.split('');

    for (let i = 0; i < count; i++) {
        let input = '';
        for (let j = 0; j < length; j++) {
            const idx = Math.floor(random() * chars.length);
            input += chars[idx];
        }
        inputs.push(input);
    }

    return inputs;
}

/**
 * Build cipher based on alphabet name
 */
function buildCipher(alphabetName, key, tweak) {
    const keyBytes = hexToBytes(key);
    const tweakBytes = hexToBytes(tweak);

    switch (alphabetName) {
        case 'digits':
            return digits(keyBytes, tweakBytes);
        case 'hex':
            return hexLower(keyBytes, tweakBytes);
        case 'base36':
            return base36Lower(keyBytes, tweakBytes);
        case 'base62':
            return base62(keyBytes, tweakBytes);
        default:
            throw new Error(`Unknown alphabet: ${alphabetName}`);
    }
}

/**
 * Get alphabet character set for input generation
 */
function getAlphabetChars(alphabetName) {
    switch (alphabetName) {
        case 'digits':
            return '0123456789';
        case 'hex':
            return '0123456789abcdef';
        case 'base36':
            return '0123456789abcdefghijklmnopqrstuvwxyz';
        case 'base62':
            return '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
        default:
            throw new Error(`Unknown alphabet: ${alphabetName}`);
    }
}

/**
 * Get radix for alphabet
 */
function getRadixForAlphabet(alphabetName) {
    const chars = getAlphabetChars(alphabetName);
    return chars.length;
}

/**
 * Run benchmark for a specific case
 */
function runBenchmark(cipher, inputs, caseType, iterations, verbose) {
    const times = [];
    let checksum = 0;

    // Precompute ciphertexts for decrypt benchmarks
    const ciphertexts = inputs.map(pt => cipher.encrypt(pt));

    for (let i = 0; i < iterations; i++) {
        const input = inputs[i % inputs.length];
        const ciphertext = ciphertexts[i % ciphertexts.length];

        const start = process.hrtime.bigint();

        let result;
        switch (caseType) {
            case 'enc':
                result = cipher.encrypt(input);
                break;
            case 'dec':
                result = cipher.decrypt(ciphertext);
                break;
            case 'roundtrip':
                const ct = cipher.encrypt(input);
                result = cipher.decrypt(ct);
                if (result !== input) {
                    throw new Error(`Round-trip verification failed: ${input} != ${result}`);
                }
                break;
        }

        const end = process.hrtime.bigint();
        times.push(Number(end - start));

        // XOR checksum to prevent dead code elimination
        for (let j = 0; j < result.length; j++) {
            checksum ^= result.charCodeAt(j);
        }

        if (verbose && (i + 1) % 10000 === 0) {
            console.error(`  Progress: ${i + 1}/${iterations}`);
        }
    }

    const totalNs = times.reduce((a, b) => a + b, 0);
    const avgNs = totalNs / times.length;
    const opsPerSec = Math.round(1e9 / avgNs);

    return {
        elapsed_ns: Math.round(totalNs),
        ns_per_op: Math.round(avgNs),
        ops_per_sec: opsPerSec,
        checksum: checksum.toString(16).padStart(8, '0')
    };
}

/**
 * Parse command-line arguments
 */
function parseArgs(args) {
    const config = { ...DEFAULT_CONFIG };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];

        switch (arg) {
            case '-h':
            case '--help':
                showHelp();
                process.exit(0);
                break;

            case '-v':
            case '--version':
                console.log(`FF3 Benchmark Tool v${VERSION}`);
                process.exit(0);
                break;

            case '--config':
                if (i + 1 >= args.length) {
                    console.error('ERR: --config requires a file path');
                    process.exit(2);
                }
                const configPath = args[++i];
                try {
                    const fileContent = fs.readFileSync(configPath, 'utf8');
                    const fileConfig = JSON.parse(fileContent);
                    Object.assign(config, fileConfig);
                } catch (err) {
                    console.error(`ERR: Failed to load config file: ${err.message}`);
                    process.exit(2);
                }
                break;

            case '--alphabet':
                if (i + 1 >= args.length) {
                    console.error('ERR: --alphabet requires a value');
                    process.exit(2);
                }
                config.alphabet = args[++i];
                break;

            case '--radix':
                if (i + 1 >= args.length) {
                    console.error('ERR: --radix requires a value');
                    process.exit(2);
                }
                config.radix = parseInt(args[++i], 10);
                break;

            case '--lengths':
                if (i + 1 >= args.length) {
                    console.error('ERR: --lengths requires a value');
                    process.exit(2);
                }
                config.lengths = parseCsvInts(args[++i]);
                break;

            case '--cases':
                if (i + 1 >= args.length) {
                    console.error('ERR: --cases requires a value');
                    process.exit(2);
                }
                config.cases = parseCsvStrings(args[++i]);
                break;

            case '--iterations':
                if (i + 1 >= args.length) {
                    console.error('ERR: --iterations requires a value');
                    process.exit(2);
                }
                config.iterations = parseInt(args[++i], 10);
                break;

            case '--warmup':
                if (i + 1 >= args.length) {
                    console.error('ERR: --warmup requires a value');
                    process.exit(2);
                }
                config.warmup = parseInt(args[++i], 10);
                break;

            case '--key':
                if (i + 1 >= args.length) {
                    console.error('ERR: --key requires a value');
                    process.exit(2);
                }
                config.key = args[++i];
                break;

            case '--tweak':
                if (i + 1 >= args.length) {
                    console.error('ERR: --tweak requires a value');
                    process.exit(2);
                }
                config.tweak = args[++i];
                break;

            case '--seed':
                if (i + 1 >= args.length) {
                    console.error('ERR: --seed requires a value');
                    process.exit(2);
                }
                config.seed = parseInt(args[++i], 10);
                break;

            case '--verbose':
                config.verbose = true;
                break;

            case '--json-out':
                if (i + 1 >= args.length) {
                    console.error('ERR: --json-out requires a file path');
                    process.exit(2);
                }
                config.jsonOut = args[++i];
                break;

            case '--quick':
                config.iterations = 10000;
                config.warmup = 1000;
                config.lengths = [9, 12];
                config.cases = ['enc', 'roundtrip'];
                break;

            default:
                console.error(`ERR: Unknown option: ${arg}`);
                showHelp();
                process.exit(1);
        }
    }

    return config;
}

/**
 * Display help information
 */
function showHelp() {
    console.log(`
FF3 Performance Benchmark Tool v${VERSION}

Usage: ff3-benchmark [options]

Options:
  --config FILE         Load configuration from JSON file
  --alphabet NAME       Alphabet to use (digits, hex, base36, base62)
  --radix N             Radix (2-62)
  --lengths CSV         Comma-separated input lengths (e.g., "9,12,16")
  --cases CSV           Comma-separated cases (enc,dec,roundtrip)
  --iterations N        Number of iterations (default: 100000)
  --warmup N            Warmup iterations (default: 10000)
  --key HEX             Encryption key (hex string)
  --tweak HEX           Tweak value (hex string)
  --seed N              Random seed for input generation
  --verbose             Show progress during benchmarks
  --json-out FILE       Write JSON results to file
  --quick               Quick benchmark (fewer iterations)
  -h, --help            Show this help message
  -v, --version         Show version

This tool measures FF3 encryption/decryption performance with different
string lengths and outputs detailed timing statistics.

⚠️  FF3 was withdrawn by NIST due to security vulnerabilities.
   This tool is for educational and research purposes only.
`);
}

/**
 * Main function
 */
function main() {
    const config = parseArgs(process.argv.slice(2));

    if (config.verbose) {
        console.error(`
🚀 FF3 Performance Benchmark Tool v${VERSION}
============================================

⚠️  FF3 was WITHDRAWN by NIST due to security vulnerabilities.
    This benchmark is for EDUCATIONAL and RESEARCH purposes only.

🔧 Configuration:
   Alphabet: ${config.alphabet}
   Radix: ${config.radix}
   Key: ${config.key.substring(0, 16)}...
   Tweak: ${config.tweak}
   Lengths: ${config.lengths.join(', ')}
   Cases: ${config.cases.join(', ')}
   Iterations: ${config.iterations}
   Warmup: ${config.warmup}
   Seed: ${config.seed}

Starting performance benchmarks...
`);
    }

    try {
        // Build cipher
        const cipher = buildCipher(config.alphabet, config.key, config.tweak);
        const alphabetChars = getAlphabetChars(config.alphabet);
        const radix = getRadixForAlphabet(config.alphabet);

        const benchmarks = [];

        // Run benchmarks for each length
        for (const length of config.lengths) {
            // Generate ring buffer of 64 varied inputs
            const inputs = generateInputs(alphabetChars, length, 64, config.seed);

            // Warmup phase
            if (config.warmup > 0) {
                if (config.verbose) {
                    console.error(`\n📊 Warmup for length ${length}...`);
                }
                runBenchmark(cipher, inputs, 'enc', config.warmup, false);
            }

            // Benchmark each case
            for (const caseType of config.cases) {
                if (config.verbose) {
                    console.error(`\n📊 Benchmarking ${caseType}_length${length}_radix${radix}...`);
                }

                const result = runBenchmark(cipher, inputs, caseType, config.iterations, config.verbose);

                benchmarks.push({
                    name: `${caseType}_length${length}_radix${radix}`,
                    n: config.iterations,
                    elapsed_ns: result.elapsed_ns,
                    ns_per_op: result.ns_per_op,
                    ops_per_sec: result.ops_per_sec,
                    checksum: result.checksum
                });

                if (config.verbose) {
                    console.error(`   ${result.ns_per_op} ns/op (${result.ops_per_sec.toLocaleString()} ops/sec)`);
                    console.error(`   Checksum: ${result.checksum}`);
                }
            }
        }

        // Build spec-compliant JSON output
        const totalDuration = benchmarks.reduce((sum, b) => sum + b.elapsed_ns, 0) / 1e9;
        const combinedChecksum = benchmarks.reduce((xor, b) => xor ^ (parseInt(b.checksum || '0', 16)), 0);

        const output = {
            metadata: {
                version: '1.0',
                timestamp: new Date().toISOString(),
                language: 'javascript',
                runtime: `Node.js ${process.version}`,
                platform: {
                    os: os.platform(),
                    arch: os.arch(),
                    cpu: os.cpus()[0]?.model || 'unknown',
                    cores: os.cpus().length
                }
            },
            configuration: {
                seed: config.seed,
                warmup_iterations: config.warmup
            },
            benchmarks: benchmarks.map(b => {
                const nameParts = b.name.split('_');
                const testCase = nameParts[0];
                const lengthStr = nameParts[1];
                const length = parseInt(lengthStr.replace('length', ''));

                return {
                    name: b.name.replace('length', 'len'),
                    test_case: testCase,
                    parameters: {
                        alphabet: config.alphabet,
                        radix: radix,
                        length: length,
                        key_bits: config.key.length * 4,
                        key_fingerprint: config.key.substring(0, 8).toUpperCase(),
                        tweak: config.tweak.toUpperCase()
                    },
                    iterations: b.n,
                    elapsed_ns: b.elapsed_ns,
                    ns_per_op: b.ns_per_op,
                    ops_per_sec: b.ops_per_sec,
                    checksum: b.checksum || '00000000'
                };
            }),
            summary: {
                total_tests: benchmarks.length,
                total_duration_sec: totalDuration,
                checksum: combinedChecksum.toString(16).padStart(8, '0')
            }
        };

        const jsonOutput = JSON.stringify(output, null, 2);

        if (config.jsonOut) {
            fs.writeFileSync(config.jsonOut, jsonOutput);
            if (config.verbose) {
                console.error(`\n✅ Results written to ${config.jsonOut}`);
            }
        } else {
            console.log(jsonOutput);
        }

        if (config.verbose) {
            console.error(`
⚠️  FF3 Format Preserving Encryption Library v${VERSION}

    FF3 was WITHDRAWN by NIST due to security vulnerabilities.
    This implementation is for EDUCATIONAL and RESEARCH purposes only.

    DO NOT use in production systems.
`);
        }

    } catch (error) {
        console.error('❌ Benchmark failed:', error.message);
        if (config.verbose) {
            console.error(error.stack);
        }
        process.exit(1);
    }
}

// Run main function
main();

File: implementations/rust/src/bin/ff3-benchmark.rs
// FF3 Performance Benchmark Tool - Spec Compliant
use fpe_ff3::{digits, hex_lower, base36_lower, base62};
use rand::{Rng, SeedableRng};
use rand::rngs::StdRng;
use serde::{Deserialize, Serialize};
use std::time::Instant;

#[derive(Serialize, Deserialize)]
#[serde(default)]
struct BenchmarkConfig {
    alphabet: String,
    radix: u32,
    lengths: Vec<usize>,
    cases: Vec<String>,
    iterations: usize,
    warmup: usize,
    key: String,
    tweak: String,
    seed: u64,
    #[serde(skip)]
    verbose: bool,
    #[serde(skip)]
    json_out: Option<String>,
}

impl Default for BenchmarkConfig {
    fn default() -> Self {
        Self {
            alphabet: "digits".to_string(),
            radix: 10,
            lengths: vec![9, 12, 16],
            cases: vec!["enc".to_string(), "dec".to_string(), "roundtrip".to_string()],
            iterations: 100000,
            warmup: 10000,
            key: "EF4359D8D580AA4F7F036D6F04FC6A94".to_string(),
            tweak: "D8E7920AFA330A73".to_string(),
            seed: 42,
            verbose: false,
            json_out: None,
        }
    }
}

#[derive(Serialize)]
struct BenchmarkParameters {
    alphabet: String,
    radix: u32,
    length: usize,
    key_bits: usize,
    key_fingerprint: String,
    tweak: String,
}

#[derive(Serialize)]
struct BenchmarkResult {
    name: String,
    test_case: String,
    parameters: BenchmarkParameters,
    iterations: usize,
    elapsed_ns: u128,
    ns_per_op: f64,
    ops_per_sec: f64,
    checksum: String,
}

#[derive(Serialize)]
struct PlatformInfo {
    os: String,
    arch: String,
    cpu: String,
    cores: usize,
}

#[derive(Serialize)]
struct MetadataInfo {
    version: String,
    timestamp: String,
    language: String,
    runtime: String,
    platform: PlatformInfo,
}

#[derive(Serialize)]
struct ConfigurationInfo {
    seed: u64,
    warmup_iterations: usize,
}

#[derive(Serialize)]
struct SummaryInfo {
    total_tests: usize,
    total_duration_sec: f64,
    checksum: String,
}

#[derive(Serialize)]
struct BenchmarkReport {
    metadata: MetadataInfo,
    configuration: ConfigurationInfo,
    benchmarks: Vec<BenchmarkResult>,
    summary: SummaryInfo,
}

fn hash_string(s: &str) -> u32 {
    let mut hash: u32 = 0;
    for byte in s.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(byte as u32);
    }
    hash
}

fn generate_inputs(count: usize, length: usize, alphabet: &str, seed: u64) -> Vec<String> {
    let mut rng = StdRng::seed_from_u64(seed);
    let mut inputs = Vec::with_capacity(count);

    let charset = match alphabet {
        "digits" => "0123456789",
        "hex" => "0123456789abcdef",
        "base36" => "0123456789abcdefghijklmnopqrstuvwxyz",
        "base62" => "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz",
        _ => "0123456789",
    };

    for _ in 0..count {
        let input: String = (0..length)
            .map(|_| {
                let idx = rng.gen_range(0..charset.len());
                charset.chars().nth(idx).unwrap()
            })
            .collect();
        inputs.push(input);
    }

    inputs
}

fn run_benchmark(
    cipher: &fpe_ff3::FF3,
    bench_case: &str,
    length: usize,
    radix: u32,
    alphabet: &str,
    iterations: usize,
    warmup: usize,
    key: &[u8],
    tweak_hex: &str,
    seed: u64,
) -> Result<BenchmarkResult, Box<dyn std::error::Error>> {
    const RING_SIZE: usize = 64;
    let inputs = generate_inputs(RING_SIZE, length, alphabet, seed);

    // Precompute ciphertexts for decrypt benchmarks
    let precomputed_cts: Vec<String> = if bench_case == "dec" {
        inputs.iter()
            .map(|pt| cipher.encrypt(pt, None).unwrap())
            .collect()
    } else {
        Vec::new()
    };

    // Warmup
    for i in 0..warmup {
        let input = &inputs[i % RING_SIZE];
        match bench_case {
            "enc" => {
                let _ = cipher.encrypt(input, None)?;
            }
            "dec" => {
                let ct = &precomputed_cts[i % RING_SIZE];
                let _ = cipher.decrypt(ct, None)?;
            }
            "roundtrip" => {
                let ct = cipher.encrypt(input, None)?;
                let _ = cipher.decrypt(&ct, None)?;
            }
            _ => {}
        }
    }

    // Measured phase
    let mut checksum: u32 = 0;
    let start = Instant::now();

    for i in 0..iterations {
        let input = &inputs[i % RING_SIZE];
        let output = match bench_case {
            "enc" => {
                let out = cipher.encrypt(input, None)?;
                checksum ^= hash_string(&out);
                out
            }
            "dec" => {
                let ct = &precomputed_cts[i % RING_SIZE];
                let out = cipher.decrypt(ct, None)?;
                checksum ^= hash_string(&out);
                out
            }
            "roundtrip" => {
                let ct = cipher.encrypt(input, None)?;
                let out = cipher.decrypt(&ct, None)?;
                checksum ^= hash_string(&out);
                out
            }
            _ => String::new(),
        };
        drop(output); // Prevent optimization
    }

    let elapsed = start.elapsed();
    let elapsed_ns = elapsed.as_nanos();
    let ns_per_op = elapsed_ns as f64 / iterations as f64;
    let ops_per_sec = 1_000_000_000.0 / ns_per_op;

    let key_fingerprint = hex::encode(&key[..std::cmp::min(4, key.len())]).to_uppercase();

    Ok(BenchmarkResult {
        name: format!("{}_len{}_radix{}", bench_case, length, radix),
        test_case: bench_case.to_string(),
        parameters: BenchmarkParameters {
            alphabet: alphabet.to_string(),
            radix,
            length,
            key_bits: key.len() * 8,
            key_fingerprint,
            tweak: tweak_hex.to_string(),
        },
        iterations,
        elapsed_ns,
        ns_per_op,
        ops_per_sec,
        checksum: format!("{:08x}", checksum),
    })
}

fn get_platform_info() -> PlatformInfo {
    PlatformInfo {
        os: std::env::consts::OS.to_string(),
        arch: std::env::consts::ARCH.to_string(),
        cpu: "unknown".to_string(),
        cores: num_cpus::get(),
    }
}

fn parse_csv_usize(s: &str) -> Vec<usize> {
    s.split(',')
        .filter_map(|p| p.trim().parse().ok())
        .collect()
}

fn parse_csv_string(s: &str) -> Vec<String> {
    s.split(',')
        .map(|p| p.trim().to_string())
        .filter(|p| !p.is_empty())
        .collect()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut config = BenchmarkConfig::default();

    // Simple arg parsing
    let args: Vec<String> = std::env::args().collect();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--help" | "-h" => {
                println!("FF3 Performance Benchmark Tool\n");
                println!("Usage: ff3-benchmark [OPTIONS]\n");
                println!("Options:");
                println!("  --config <file>        JSON config file");
                println!("  --alphabet <name>      Alphabet: digits|hex|base36|base62 (default: digits)");
                println!("  --radix <n>            Radix 2-62 (default: 10)");
                println!("  --lengths <csv>        Comma-separated lengths (default: 9,12,16)");
                println!("  --cases <csv>          Test cases: enc,dec,roundtrip (default: all)");
                println!("  --iterations <n>       Measured iterations (default: 100000)");
                println!("  --warmup <n>           Warmup iterations (default: 10000)");
                println!("  --key <hex>            Hex-encoded key");
                println!("  --tweak <hex>          Hex-encoded tweak");
                println!("  --seed <n>             Random seed (default: 42)");
                println!("  --quick                Reduce iterations 10x");
                println!("  --json-out <file>      Write JSON results to file");
                println!("  --verbose              Show progress messages");
                println!("  -h, --help             Show this help");
                return Ok(());
            }
            "--config" => {
                if i + 1 < args.len() {
                    let file_content = std::fs::read_to_string(&args[i + 1])?;
                    config = serde_json::from_str(&file_content)?;
                    i += 1;
                }
            }
            "--alphabet" => {
                if i + 1 < args.len() {
                    config.alphabet = args[i + 1].clone();
                    i += 1;
                }
            }
            "--radix" => {
                if i + 1 < args.len() {
                    config.radix = args[i + 1].parse()?;
                    i += 1;
                }
            }
            "--lengths" => {
                if i + 1 < args.len() {
                    config.lengths = parse_csv_usize(&args[i + 1]);
                    i += 1;
                }
            }
            "--cases" => {
                if i + 1 < args.len() {
                    config.cases = parse_csv_string(&args[i + 1]);
                    i += 1;
                }
            }
            "--iterations" => {
                if i + 1 < args.len() {
                    config.iterations = args[i + 1].parse()?;
                    i += 1;
                }
            }
            "--warmup" => {
                if i + 1 < args.len() {
                    config.warmup = args[i + 1].parse()?;
                    i += 1;
                }
            }
            "--key" => {
                if i + 1 < args.len() {
                    config.key = args[i + 1].clone();
                    i += 1;
                }
            }
            "--tweak" => {
                if i + 1 < args.len() {
                    config.tweak = args[i + 1].clone();
                    i += 1;
                }
            }
            "--seed" => {
                if i + 1 < args.len() {
                    config.seed = args[i + 1].parse()?;
                    i += 1;
                }
            }
            "--quick" => {
                config.iterations /= 10;
                config.warmup /= 10;
            }
            "--json-out" => {
                if i + 1 < args.len() {
                    config.json_out = Some(args[i + 1].clone());
                    i += 1;
                }
            }
            "--verbose" => {
                config.verbose = true;
            }
            _ => {}
        }
        i += 1;
    }

    let key = hex::decode(&config.key)?;
    let tweak = hex::decode(&config.tweak)?;

    // Build cipher based on alphabet
    let cipher = match config.alphabet.as_str() {
        "digits" => digits(&key, &tweak)?,
        "hex" => hex_lower(&key, &tweak)?,
        "base36" => base36_lower(&key, &tweak)?,
        "base62" => base62(&key, &tweak)?,
        _ => digits(&key, &tweak)?,
    };

    let total_start = Instant::now();
    let mut benchmarks = Vec::new();
    let mut overall_checksum: u32 = 0;

    for length in &config.lengths {
        for bench_case in &config.cases {
            if config.verbose {
                eprintln!("Running {}_len{}_radix{}...", bench_case, length, config.radix);
            }

            let result = run_benchmark(
                &cipher,
                bench_case,
                *length,
                config.radix,
                &config.alphabet,
                config.iterations,
                config.warmup,
                &key,
                &config.tweak,
                config.seed,
            )?;

            overall_checksum ^= u32::from_str_radix(&result.checksum, 16)?;
            benchmarks.push(result);
        }
    }

    let total_duration = total_start.elapsed();

    let report = BenchmarkReport {
        metadata: MetadataInfo {
            version: "1.0".to_string(),
            timestamp: chrono::Utc::now().to_rfc3339(),
            language: "rust".to_string(),
            runtime: "rustc".to_string(),
            platform: get_platform_info(),
        },
        configuration: ConfigurationInfo {
            seed: config.seed,
            warmup_iterations: config.warmup,
        },
        benchmarks,
        summary: SummaryInfo {
            total_tests: config.lengths.len() * config.cases.len(),
            total_duration_sec: total_duration.as_secs_f64(),
            checksum: format!("{:08x}", overall_checksum),
        },
    };

    let json = serde_json::to_string_pretty(&report)?;

    if let Some(path) = config.json_out {
        std::fs::write(path, json)?;
    } else {
        println!("{}", json);
    }

    Ok(())
}
